File: go/schemas.go
package confidentagent

import "encoding/json"

type ResponseMode string

const (
	ResponseModeHTTPResponse  ResponseMode = "HTTP_RESPONSE"
	ResponseModeSSEStreaming  ResponseMode = "SSE_STREAMING"
	ResponseModeHTTPStreaming ResponseMode = "HTTP_STREAMING"
)

type RelayRequest struct {
	ID           string            `json:"id"`
	Method       string            `json:"method"`
	URL          string            `json:"url"`
	Headers      map[string]string `json:"headers"`
	Body         any               `json:"body"`
	Timeout      int               `json:"timeout"`
	ResponseMode ResponseMode      `json:"response_mode"`
}

func (r *RelayRequest) UnmarshalJSON(data []byte) error {
	type relayRequest RelayRequest
	var raw relayRequest
	if err := json.Unmarshal(data, &raw); err != nil {
		return err
	}
	if raw.Headers == nil {
		raw.Headers = map[string]string{}
	}
	if raw.Timeout == 0 {
		raw.Timeout = 60
	}
	if raw.ResponseMode == "" {
		raw.ResponseMode = ResponseModeHTTPResponse
	}
	*r = RelayRequest(raw)
	return nil
}

type RelayResponseType string

const (
	RelayResponseTypeResponse  RelayResponseType = "response"
	RelayResponseTypeError     RelayResponseType = "error"
	RelayResponseTypeChunk     RelayResponseType = "chunk"
	RelayResponseTypeStreamEnd RelayResponseType = "stream_end"
)

type RelayResponse struct {
	ID         string            `json:"id"`
	Type       RelayResponseType `json:"type"`
	StatusCode int               `json:"statusCode"`
	Headers    map[string]string `json:"headers"`
	Body       string            `json:"body"`
}

func (r RelayResponse) MarshalJSON() ([]byte, error) {
	type relayResponse RelayResponse
	if r.Headers == nil {
		r.Headers = map[string]string{}
	}
	return json.Marshal(relayResponse(r))
}

type ToolCall struct {
	Name            string         `json:"name"`
	Description     *string        `json:"description"`
	InputParameters map[string]any `json:"input_parameters"`
	Output          any            `json:"output"`
}

type Turn struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type Request struct {
	Input            string         `json:"input"`
	Context          []string       `json:"context"`
	RetrievalContext []string       `json:"retrieval_context"`
	ExpectedOutput   *string        `json:"expected_output"`
	Turns            []Turn         `json:"turns"`
	Scenario         *string        `json:"scenario"`
	State            any            `json:"state"`
	Prompts          map[string]any `json:"prompts"`
	Hyperparameters  map[string]any `json:"hyperparameters"`
	TestCaseID       *string        `json:"test_case_id"`
	TurnID           *string        `json:"turn_id"`
}

type Response struct {
	Output           string     `json:"output"`
	RetrievalContext []string   `json:"retrieval_context"`
	ToolsCalled      []ToolCall `json:"tools_called"`
	State            any        `json:"state"`
}

File: typescript/src/schemas.ts
export enum ResponseMode {
  HTTP_RESPONSE = "HTTP_RESPONSE",
  SSE_STREAMING = "SSE_STREAMING",
  HTTP_STREAMING = "HTTP_STREAMING",
}

export interface RelayRequest {
  id: string;
  method: string;
  url: string;
  headers: Record<string, string>;
  body: unknown;
  timeout: number;
  responseMode: ResponseMode;
}

export enum RelayResponseType {
  RESPONSE = "response",
  ERROR = "error",
  CHUNK = "chunk",
  STREAM_END = "stream_end",
}

export interface RelayResponse {
  id: string;
  type: RelayResponseType;
  statusCode: number;
  headers: Record<string, string>;
  body: string;
}

export interface ToolCall {
  [key: string]: unknown;
  name: string;
  description?: string | null;
  inputParameters?: Record<string, unknown> | null;
  output?: unknown;
}

export interface Turn {
  [key: string]: unknown;
  role: string;
  content: string;
}

export interface Request {
  [key: string]: unknown;
  input: string;
  context: string[];
  retrievalContext: string[];
  expectedOutput: string | null;
  turns: Turn[];
  scenario: string | null;
  state: unknown;
  prompts: Record<string, unknown> | null;
  hyperparameters: Record<string, unknown> | null;
  testCaseId: string | null;
  turnId: string | null;
}

export interface Response {
  output: string;
  retrievalContext?: string[] | null;
  toolsCalled?: ToolCall[] | null;
  state?: unknown;
}

export function parseRelayRequest(raw: string): RelayRequest {
  const data = JSON.parse(raw) as Record<string, unknown>;
  if (
    typeof data.id !== "string" ||
    typeof data.method !== "string" ||
    typeof data.url !== "string"
  ) {
    throw new Error("id, method, and url are required strings");
  }
  return {
    id: data.id,
    method: data.method,
    url: data.url,
    headers: (data.headers as Record<string, string> | undefined) ?? {},
    body: data.body ?? null,
    timeout: (data.timeout as number | undefined) ?? 60,
    responseMode:
      (data.response_mode as ResponseMode | undefined) ?? ResponseMode.HTTP_RESPONSE,
  };
}

export function parseRequest(body: Record<string, unknown>): Request {
  const {
    input = "",
    context = [],
    retrieval_context: retrievalContext = [],
    expected_output: expectedOutput = null,
    turns = [],
    scenario = null,
    state = null,
    prompts = null,
    hyperparameters = null,
    test_case_id: testCaseId = null,
    turn_id: turnId = null,
    ...extra
  } = body;
  return {
    ...extra,
    input: input as string,
    context: context as string[],
    retrievalContext: retrievalContext as string[],
    expectedOutput: expectedOutput as string | null,
    turns: turns as Turn[],
    scenario: scenario as string | null,
    state,
    prompts: prompts as Record<string, unknown> | null,
    hyperparameters: hyperparameters as Record<string, unknown> | null,
    testCaseId: testCaseId as string | null,
    turnId: turnId as string | null,
  };
}

export function serializeResponse(response: Response): string {
  return JSON.stringify({
    output: response.output,
    retrieval_context: response.retrievalContext ?? null,
    tools_called: response.toolsCalled
      ? response.toolsCalled.map(serializeToolCall)
      : null,
    state: response.state ?? null,
  });
}

function serializeToolCall(toolCall: ToolCall): Record<string, unknown> {
  const { name, description, inputParameters, output, ...extra } = toolCall;
  return {
    name,
    description: description ?? null,
    input_parameters: inputParameters ?? null,
    output: output ?? null,
    ...extra,
  };
}
